### PR TITLE
Add a check to prevent call function on a boolean when category and context tables are empty. fixes #159

### LINKED ISSUE #159: Error on admin/sonata/classification/category/tree when database is empty
Hello,

I encountered an issue in dev master version of SonataClassificationBundle :

```
Call to a member function getContext() on boolean in 
[...]/SonataClassificationBundle/Controller/CategoryAdminController.php on line 80
```

I check the issue list but I do not see it, so I report here the bug. This bug only happen when database is empty. The schema is ok, but there is voluntarily no category and no context in my database. 

If you want to check, this is my composer.json file :

```
"require": {
        "php": ">=5.3.9",
        "symfony/symfony": "2.7.*",
        "doctrine/orm": "^2.4.8",
        "doctrine/doctrine-bundle": "~1.4",
        "doctrine/doctrine-fixtures-bundle": "dev-master",
        "doctrine/migrations": "~1",
        "doctrine/doctrine-migrations-bundle": "~1",
        "symfony/assetic-bundle": "~2.3",
        "symfony/swiftmailer-bundle": "~2.3",
        "symfony/monolog-bundle": "~2.4",
        "sensio/distribution-bundle": "~4.0",
        "sensio/framework-extra-bundle": "^3.0.2",
        "incenteev/composer-parameter-handler": "~2.0",
        "friendsofsymfony/user-bundle": "~1.3",
        "friendsofsymfony/message-bundle": "^1.2",
        "sonata-project/admin-bundle": "^2.4@dev",
        "sonata-project/block-bundle": "~2.3",
        "sonata-project/classification-bundle": "dev-master",
        "sonata-project/datagrid-bundle": "dev-master",
        "sonata-project/doctrine-orm-admin-bundle": "^2.4@dev",
        "sonata-project/easy-extends-bundle": "^2.1",
        "sonata-project/exporter": "^1.4",
        "sonata-project/intl-bundle": "^2.2",
        "sonata-project/user-bundle": "dev-master",
        "knplabs/knp-menu-bundle": "~2.0",
        "knplabs/knp-menu": "~2.0"
    },
```

Here is more information about my project and my server :
- https://github.com/Alexandre-T/jeu-de-role
- Server Apache 2.4
- PHP 5.6.13-0+deb8u1
- Debian Jessie

I **already developed** an unit test case which fails on that bug and I **already fixed it**, I will push a request in a few minutes, I have to to read your contribution rules before that.**

Sorry for my english, I am not very fluent.

### PR DESCRIPTION
**Pull request by Alexandre-T**
(Sorry, not fluent in English)

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  |  no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #159 
| License       | MIT

If you want to see the bug, use the ``CategoryAdminControllerTest`` to the master version and launch phpunit. I have done two modifications, the first on ``CategoryAdminController`` and the second to ``tree.html.twig``. My Controller Test only test the Controller. I don't know how to realise a unit test on ``tree.html.twig``. 

This is my first participation to a sonata bundle. I hope to do it well. I read the howto [make a pull request](http://symfony.com/doc/current/contributing/code/patches.html#make-a-pull-request) . I read the [guidelines for contributing](https://github.com/sonata-project/SonataClassificationBundle/blob/master/CONTRIBUTING.md) too.

* *You add valid test cases.* **Checked**
* *Tests are green.* **Checked**
* *The related documentation is up-to-date.* No change to documentation
* *You make the PR on the same branch you based your changes on. If you see commits that you did not make in your PR, you're doing it wrong.* I make change on fix-issue-159 branch as ask in [make a pull request](http://symfony.com/doc/current/contributing/code/patches.html#make-a-pull-request)
* *Also don't forget to add a comment when you update a PR with a ping to the maintainer (@username), so he/she will get a notification.* ping @rande 

Hope to help you, and thanks for this awesome project !

### CODE REVIEW

**Comment by OskarStark** [comment]
Hey @Alexandre-T 

first of all, thank you for your excellent contribution.
Can you please explain the bug in a few sentences?

And after that please squash your commits and and use a clear commit message of WHAT your fixing

Thank you :+1:

**Comment by Alexandre-T** [comment]
Hey @OskarStark, 

Thank you for your reply,

I explained the issue here : #159 . When categorie and context tables are empty, the tree view can not be shown to user. There is a bug.  Line 80 application try to call function on $mainCategory. There is no category so $mainCategory is a boolean (False) instead of an Category entity. 

```Call to a member function getContext() on boolean in 
[...]/SonataClassificationBundle/Controller/CategoryAdminController.php on line 80
```

There is a similar problem in the tree.html.twig so i added a to verifiy that element is not empty

I will squash my commits (I will learn to do it before ;-) ). 

About the commit message, what do you think about this text : 

>Add a test to prevent call function on a boolean when category and context tables are empty.
>Fixed issue 159#

**Comment by OskarStark** [comment]
thank you very much @Alexandre-T 

:+1: for the commit message, but use ``... fixes #159`` at the end please

For squashing you can have a look here:
https://github.com/sonata-project/SonataAdminBundle/blob/master/CONTRIBUTING.md

**Comment by Alexandre-T** [comment]
Thank you for the link, @OskarStark 

Jobs done ! 

Timeout for one of the travis tests.

In .travis.yml file, you could perhaps change : 

`travis_wait composer update --prefer-dist --no-interaction $COMPOSER_FLAGS`

by 

`travis_wait 40 composer update --prefer-dist --no-interaction $COMPOSER_FLAGS`

**Comment by OskarStark** [comment]
ping @Soullivaneuh can you please help us with the composer stuff and have a look at this PR?

Thank you!

**Comment by OskarStark** [comment]
@Alexandre-T i changed the headline of this PR, can you please use exactly this as the commit message?

please use ``#159`` instead of ``159#``, as i'm not sure if github would parse it correctly

thank you

**Comment by Alexandre-T** [comment]
Hey @OskarStark,

I changed it, sorry for my english ;-)

**Comment by OskarStark** [comment]
no problem :)

**Comment by sagikazarmark** [comment]
I don't think adding higher timout to travis wait is a real solution.

File: Controller/CategoryAdminController.php [hunk 1]
```diff
@@ -74,7 +75,7 @@ public function treeAction(Request $request)
 
         $rootCategories = $categoryManager->getRootCategories(false);
 
-        if (!$currentContext) {
+        if (!$currentContext && $rootCategories) {
```
Comment (sagikazarmark) [inline Controller/CategoryAdminController.php]: I think it is not a boolean value. It is probably an array or null, guessing from the function name. A proper emptyness check would be better I think.

File: Resources/views/CategoryAdmin/tree.html.twig [hunk 2]
```diff
@@ -19,20 +19,22 @@ file that was distributed with this source code.
...
-            <li class="sonata-ba-list-field" objectId="{{ element.id }}" >
-                <div class="sonata-tree__item"{% if depth < 2 %} data-treeview-toggled{% endif %}>
-                    {% if element.parent or root %}<i class="fa fa-caret-right" data-treeview-toggler></i>{% endif %}
-                    <a class="sonata-tree__item__edit" href="{{ admin.generateObjectUrl('edit', element) }}">{{ element.name }}</a>
-                    <i class="text-muted">{{ element.description }}</i>
-                    {#<a class="label label-default pull-right" href="{{ admin.generateObjectUrl('edit', element) }}">edit <i class="fa fa-magic"></i></a>#}
-                    {% if element.enabled %}<span class="label label-success pull-right"><i class="fa fa-check"></i> {{ admin.trans('active', {}, admin.translationDomain) }}</span>{% endif %}
-                    {% if not element.enabled %}<span class="label label-danger pull-right"><i class="fa fa-times"> {{ admin.trans('disabled', {}, admin.translationDomain) }}</i></span>{% endif %}
-                </div>
+            {% if element %}
```
Comment (sagikazarmark) [inline Resources/views/CategoryAdmin/tree.html.twig]: What is this check for? Emptyness check?

**Comment by sagikazarmark** [comment]
:+1:

File: Controller/CategoryAdminController.php (hunk 1 above)
Comment (Alexandre-T) [inline Controller/CategoryAdminController.php]: Yes, it is, I can change this line by :
```if (!$currentContext && !empty($rootCategories))```
or 
```if (!($currentContext || empty($rootCategories)))```

File: Resources/views/CategoryAdmin/tree.html.twig (hunk 2 above)
Comment (Alexandre-T) [inline Resources/views/CategoryAdmin/tree.html.twig]: Yes, it is an emptyness check. When element is empty, element.id call crash the view generation.  I am not very fluent with Twig. When I saw this code, I'm asking myself why it entered in the foreach loop, even if element is empty.

**Comment by OskarStark** [comment]
Restarted Travis

**Comment by Alexandre-T** [comment]
@OskarStark Do you want that I update my condition like @sagikazarmark suggests it in the CategoryAdminController.php ?

File: Controller/CategoryAdminController.php (hunk 1 above)
Comment (OskarStark) [inline Controller/CategoryAdminController.php]: :+1:

**Comment by Alexandre-T** [comment]
@OskarStark Condition updated like  @sagikazarmark had suggested it.
But all tests are crashing. I think that this is like  [this issue](https://github.com/sonata-project/SonataAdminBundle/issues/3459)

File: Controller/CategoryAdminController.php [hunk 3]
```diff
@@ -74,7 +75,7 @@ public function treeAction(Request $request)
 
         $rootCategories = $categoryManager->getRootCategories(false);
 
-        if (!$currentContext) {
+        if (!($currentContext || empty($rootCategories))) {
```
Comment (core23) [inline Controller/CategoryAdminController.php]: Looks wrong to me. Please remove the breaks around this condition

File: Resources/views/CategoryAdmin/tree.html.twig [hunk 4]
```diff
@@ -19,29 +19,31 @@ file that was distributed with this source code.
...
 
 {% block tab_menu %}
     {% include 'SonataClassificationBundle:CategoryAdmin:list_tab_menu.html.twig' with {
-        'mode':   'tree',
-        'action': action,
-        'admin':  admin,
+    'mode':   'tree',
+    'action': action,
+    'admin':  admin,
     } only %}
```
Comment (core23) [inline Resources/views/CategoryAdmin/tree.html.twig]: Please leave idention as it was

**Comment by OskarStark** [comment]
@Alexandre-T please rebase to get the latest travis conf

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.
